state/machine_upgradeseries_test.go

```diff
@@ -0,0 +1,202 @@
+	AssertMachineIsNOTLockedForPrepare(c, mach)
+
+	err = mach.CreateUpgradeSeriesLock([]string{}, "xenial")
+	c.Assert(err, jc.ErrorIsNil)
+	AssertMachineLockedForPrepare(c, mach)
+
+	err = mach.RemoveUpgradeSeriesLock()
+	c.Assert(err, jc.ErrorIsNil)
+	AssertMachineIsNOTLockedForPrepare(c, mach)
+}
+
+func (s *MachineSuite) TestForceMarksSeriesLockUnlocksMachineForCleanup(c *gc.C) {
+	mach, err := s.State.AddMachine("precise", state.JobHostUnits)
+	c.Assert(err, jc.ErrorIsNil)
+	AssertMachineIsNOTLockedForPrepare(c, mach)
+
+	err = mach.CreateUpgradeSeriesLock([]string{}, "xenial")
+	c.Assert(err, jc.ErrorIsNil)
+	AssertMachineLockedForPrepare(c, mach)
+
+	err = mach.ForceDestroy()
+	c.Assert(err, jc.ErrorIsNil)
+
+	// After a forced destroy an upgrade series lock on a machine should be
+	// marked for cleanup and therefore should be cleaned up if anything
+	// should trigger a state cleanup.
+	s.State.Cleanup()
+
+	// The machine, since it was destroyed, its lock should have been
+	// cleaned up. Checking to see if the machine is not locked, that is,
+	// checking to see if no lock exist for the machine should yield a
+	// positive result.
+	AssertMachineIsNOTLockedForPrepare(c, mach)
+}
+
+func (s *MachineSuite) TestCompleteSeriesUpgradeShouldFailWhenMachineIsNotComplete(c *gc.C) {
+	err := s.machine.CreateUpgradeSeriesLock([]string{}, "cosmic")
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = s.machine.CompleteUpgradeSeries()
+	assertMachineIsNotReadyForCompletion(c, err)
+}
+
+func (s *MachineSuite) TestCompleteSeriesUpgradeShouldSucceedWhenMachinePrepareIsComplete(c *gc.C) {
+	unit0 := s.addMachineUnit(c, s.machine)
+	err := s.machine.CreateUpgradeSeriesLock([]string{unit0.Name()}, "cosmic")
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = s.machine.SetMachineUpgradeSeriesStatus(model.PrepareCompleted)
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = s.machine.CompleteUpgradeSeries()
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *MachineSuite) TestCompleteSeriesUpgradeShouldSetCompleteStatusOfMachine(c *gc.C) {
+	err := s.machine.CreateUpgradeSeriesLock([]string{}, "cosmic")
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = s.machine.SetMachineUpgradeSeriesStatus(model.PrepareCompleted)
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = s.machine.CompleteUpgradeSeries()
+	c.Assert(err, jc.ErrorIsNil)
+
+	sts, err := s.machine.MachineUpgradeSeriesStatus()
+	c.Assert(err, jc.ErrorIsNil)
+
+	c.Assert(sts, gc.Equals, model.CompleteStarted)
+}
+
+func (s *MachineSuite) TestCompleteSeriesUpgradeShouldFailIfAlreadyInCompleteState(c *gc.C) {
+	unit0 := s.addMachineUnit(c, s.machine)
+	err := s.machine.CreateUpgradeSeriesLock([]string{unit0.Name()}, "cosmic")
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = s.machine.SetMachineUpgradeSeriesStatus(model.PrepareCompleted)
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = s.machine.CompleteUpgradeSeries()
+	c.Assert(err, jc.ErrorIsNil)
+
+	c.Skip("Waiting for machine status updates. This test will not pass unit machine updates its status.")
```

Comment:
Heads up. This will be fixed when you merge with develop. The fix was in `manadarts_upgrade_series_worker`.

Reply:
Thanks. Unskipped and passing.